File: frontend/src/app/services/prestador.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Prestador } from '../models/prestador.model';
import { Categoria } from '../models/categoria.enum';
import { StatusPrestador } from '../models/prestador.model';

@Injectable({
  providedIn: 'root'
})
export class PrestadorService {

  private apiUrl = 'http://localhost:8080/api/prestadores';

  constructor(private http: HttpClient) { }

  listarTodos(): Observable<Prestador[]> {
    return this.http.get<Prestador[]>(this.apiUrl);
  }

  buscarPorId(id: number): Observable<Prestador> {
    return this.http.get<Prestador>(`${this.apiUrl}/${id}`);
  }

  criar(prestador: Prestador): Observable<Prestador> {
    return this.http.post<Prestador>(this.apiUrl, prestador);
  }

  atualizar(id: number, prestador: Prestador): Observable<Prestador> {
    return this.http.put<Prestador>(`${this.apiUrl}/${id}`, prestador);
  }

  deletar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  filtrarPorEspecialidade(especialidade: Categoria): Observable<Prestador[]> {
    return this.http.get<Prestador[]>(`${this.apiUrl}/especialidade/${especialidade}`);
  }

  filtrarPorStatus(status: StatusPrestador): Observable<Prestador[]> {
    return this.http.get<Prestador[]>(`${this.apiUrl}/status/${status}`);
  }

  listarDisponiveis(especialidade: Categoria): Observable<Prestador[]> {
    return this.http.get<Prestador[]>(`${this.apiUrl}/disponiveis/${especialidade}`);
  }
}

File: backend/src/main/java/com/lucas/chamados/service/PrestadorService.java
package com.lucas.chamados.service;

import com.lucas.chamados.model.Prestador;
import com.lucas.chamados.model.Categoria;
import com.lucas.chamados.model.StatusPrestador;
import com.lucas.chamados.repository.PrestadorRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PrestadorService {

    @Autowired
    private PrestadorRepository repository;

    public List<Prestador> listarTodos() {
        return repository.findAll();
    }

    public Optional<Prestador> buscarPorId(Long id) {
        return repository.findById(id);
    }

    public Prestador criar(Prestador prestador) {
        return repository.save(prestador);
    }

    public Prestador atualizar(Long id, Prestador prestadorAtualizado) {
        Prestador prestador = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Prestador não encontrado"));

        prestador.setNome(prestadorAtualizado.getNome());
        prestador.setEmpresa(prestadorAtualizado.getEmpresa());
        prestador.setCnpj(prestadorAtualizado.getCnpj());
        prestador.setTelefone(prestadorAtualizado.getTelefone());
        prestador.setEmail(prestadorAtualizado.getEmail());
        prestador.setEspecialidade(prestadorAtualizado.getEspecialidade());
        prestador.setStatus(prestadorAtualizado.getStatus());
        prestador.setObservacoes(prestadorAtualizado.getObservacoes());

        return repository.save(prestador);
    }

    public void deletar(Long id) {
        repository.deleteById(id);
    }

    public List<Prestador> filtrarPorEspecialidade(Categoria especialidade) {
        return repository.findByEspecialidade(especialidade);
    }

    public List<Prestador> filtrarPorStatus(StatusPrestador status) {
        return repository.findByStatus(status);
    }

    public List<Prestador> listarDisponiveisPorEspecialidade(Categoria especialidade) {
        return repository.findByEspecialidadeAndStatus(especialidade, StatusPrestador.DISPONIVEL);
    }
}